protocol/src/model/packets.rs

```diff
@@ -84,6 +84,48 @@ impl SerializedPacket {
                 Some(format!("packet_{:#x}.bin", self.id.0)),
             ))
     }
+
+    pub fn serialize_compressing(&self, compression: Option<usize>) -> Result<Bytes, Error> {
+        if compression.filter(|x| self.length >= *x).is_some() {
+            let data_length = self.id.length_of() + self.data.len();
+            let datalength = VarInt::<i32>(data_length.try_into().unwrap());
+            let length =
+                datalength.length_of() + Compress((&self.id, &self.data), Zlib).serialize()?.len();
+            let pack = SerializedPacketCompressed {
+                length,
+                data_length,
+                id: self.id,
+                data: self.data.clone(),
+            };
+            pack.serialize()
+        } else {
+            self.serialize()
+        }
+    }
+
+    pub fn deserialize_compressing<'a>(
+        compression: Option<usize>,
+    ) -> pfn_type!(&'a [u8], Self, Extra<<Self as Deserialize>::Context>) {
+        move |input| {
+            if let Some(cmp) = compression {
```

Comment:
the datalength could be 0 just because the client / server decided so. it shouldn't happen, but it should be accounted for.

Reply:
> shouldn't the function check if the packet's uncompressed length field is 0 (meaning the packet being deserialized is uncompressed) before trying to decompress the packet?

this all happens when deserializing, like validation after deserialization, if that makes sense

Reply:
> > shouldn't the function check if the packet's uncompressed length field is 0 (meaning the packet being deserialized is uncompressed) before trying to decompress the packet?
> 
> this all happens when deserializing, like validation after deserialization, if that makes sense

Neither `SerializedPacket::deserialize_compressing` nor `SerializedPacketCompressed::deserialize` seem to handle the case I mentioned (where compression is enabled but the packet is uncompressed).


> the datalength could be 0 just because the client / server decided so. it shouldn't happen, but it should be accounted for.

`datalength = 0` is how the sender should indicate to the receiver that the packet is uncompressed when compression is enabled. AFAIK if the packet is uncompressed but it's datalength isn't 0 or if the packet is compressed but it's datalength is 0 then the sender has a bug, and I think in that case it should be up to the caller of the function to handle the error (for example if it's the server it should disconnect the client that sent the broken packet).